src/gpuarray_buffer_opencl.c

```diff
@@ -113,6 +172,27 @@ cl_ctx *cl_make_ctx(cl_context ctx, int flags) {
     return NULL;
   }
   res->refcnt--; /* Prevent ref loop */
+
+  /* Create per-context OpenCL preamble */
+
+  // Create a dummy kernel and check GA_KERNEL_PROP_PREFLSIZE
+  rlk[0] = dummy_kern;
+  len = sizeof(dummy_kern);
+  // this dummy kernel does not require a CLUDA preamble
+  m = cl_newkernel((gpucontext *)res, 1, rlk, &len, "kdummy", 0, NULL, 0, &ret, NULL);
+  assert((m != NULL) && "Cannot create a dummy OpenCL kernel!");
+  ret = cl_property((gpucontext *)res, NULL, m, GA_KERNEL_PROP_PREFLSIZE, &warp_size);
+  if (ret != GA_NO_ERROR) warp_size = 128;
```

Comment:
I think it would be better to return an error if we can't determine the
warp size. Normally, we should always have that information and if we use
the wrong value, we will probably return bad result. So an error here is
better I think.

On Thu, Jul 14, 2016 at 12:43 PM, abergeron <notifications@github.com>
wrote:

> In src/gpuarray_buffer_opencl.c
> <https://github.com/Theano/libgpuarray/pull/217#discussion_r70839798>:
>
> > @@ -113,6 +172,27 @@ cl_ctx *cl_make_ctx(cl_context ctx, int flags) {
> >      return NULL;
> >    }
> >    res->refcnt--; /* Prevent ref loop */
> > +
> > +  /* Create per-context OpenCL preamble */
> > +
> > +  // Create a dummy kernel and check GA_KERNEL_PROP_PREFLSIZE
> > +  rlk[0] = dummy_kern;
> > +  len = sizeof(dummy_kern);
> > +  // this dummy kernel does not require a CLUDA preamble
> > +  m = cl_newkernel((gpucontext *)res, 1, rlk, &len, "kdummy", 0, NULL, 0, &ret, NULL);
> > +  assert((m != NULL) && "Cannot create a dummy OpenCL kernel!");
> > +  ret = cl_property((gpucontext *)res, NULL, m, GA_KERNEL_PROP_PREFLSIZE, &warp_size);
> > +  if (ret != GA_NO_ERROR) warp_size = 128;
>
> If there is an error assume the warp size is 64, not 128.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/libgpuarray/pull/217/files/a80391e0f30dd8b0d5fab0b147f16d0f7003708e..cb74846510ce3d16bc9db110de9e5eb5caa1b81b#r70839798>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AALC-_e3ZMknnrjxE9cF4NTdbmesXwGlks5qVmcsgaJpZM4JLod8>
> .
>